<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 4 Starting with data | Data manipulation and visualisation with R" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Course material for the Data manipulation and visualisation with R training (SMCS &amp; CBIO, UCLouvain)." />


<meta name="author" content="Laurent Gatto, Manon Martin &amp; Axelle Loriot" />

<meta name="date" content="2021-09-24" />


<meta name="description" content="Course material for the Data manipulation and visualisation with R training (SMCS &amp; CBIO, UCLouvain).">

<title>Chapter 4 Starting with data | Data manipulation and visualisation with R</title>

<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Data manipulation and visualisation with R<p><p class="author">Laurent Gatto, Manon Martin & Axelle Loriot</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html">Preamble</a>
<a href="sec-dataorg.html"><span class="toc-section-number">1</span> Data organisation with Spreadsheets</a>
<a href="sec-rrstudio.html"><span class="toc-section-number">2</span> R and RStudio</a>
<a href="sec-startr.html"><span class="toc-section-number">3</span> Introduction to R</a>
<a id="active-page" href="sec-startdata.html"><span class="toc-section-number">4</span> Starting with data</a><ul class="toc-sections">
<li class="toc"><a href="#presentation-of-the-gene-expression-data"> Presentation of the gene expression data</a></li>
<li class="toc"><a href="#what-are-data-frames"> What are data frames?</a></li>
<li class="toc"><a href="#NA"> Inspecting <code>data.frame</code> Objects</a></li>
<li class="toc"><a href="#indexing-and-subsetting-data-frames"> Indexing and subsetting data frames</a></li>
<li class="toc"><a href="#factors"> Factors</a></li>
<li class="toc"><a href="#matrices"> Matrices</a></li>
<li class="toc"><a href="#summary-of-r-objects"> Summary of R objects</a></li>
<li class="toc"><a href="#lists"> Lists</a></li>
<li class="toc"><a href="#sec-exportandsave"> Exporting and saving data</a></li>
</ul>
<a href="sec-dplyr.html"><span class="toc-section-number">5</span> Manipulating and analyzing data with dplyr</a>
<a href="sec-vis.html"><span class="toc-section-number">6</span> Data visualization</a>
<a href="sec-join.html"><span class="toc-section-number">7</span> Joining tables</a>
<a href="sec-next.html"><span class="toc-section-number">8</span> Next steps</a>
<a href="session-information.html"><span class="toc-section-number">9</span> Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><div id="sec-startdata" class="section level1" number="4">
<h1>
<span class="header-section-number">Chapter 4</span> Starting with data</h1>
<p><strong>Learning Objectives</strong></p>
<ul>
<li>Describe what a data frame is.</li>
<li>Load external data from a .csv file into a data frame.</li>
<li>Summarize the contents of a data frame.</li>
<li>Describe what a factor is.</li>
<li>Convert between strings and factors.</li>
<li>Reorder and rename factors.</li>
<li>Format dates.</li>
<li>Export and save data.</li>
</ul>
<div id="presentation-of-the-gene-expression-data" class="section level2" number="4.1">
<h2>
<span class="header-section-number">4.1</span> Presentation of the gene expression data<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('presentation-of-the-gene-expression-data')" onmouseout="reset_tooltip('presentation-of-the-gene-expression-data-tooltip')"><span class="tooltiptext" id="presentation-of-the-gene-expression-data-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>We are going to use part of the data published by <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5544260/">Blackmore <em>et al.</em>
(2017)</a>, <em>The
effect of upper-respiratory infection on transcriptomic changes in the
CNS</em>. The goal of the study was to determine the effect of an
upper-respiratory infection on changes in RNA transcription occuring
in the cerebellum and spinal cord post infection. Gender matched eight
week old C57BL/6 mice were inoculated saline or with Influenza A by
intranasal route and transcriptomic changes in the cerebellum and
spinal cord tissues were evaluated by RNA-seq at days 0
(non-infected), 4 and 8.</p>
<p>The dataset is stored as a comma separated value (CSV) file. Each row
holds information for a single RNA expression measurement, and the
columns represent:</p>
<table>
<colgroup>
<col width="11%">
<col width="88%">
</colgroup>
<thead><tr class="header">
<th>Column</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>gene</td>
<td>The name of the gene that was measured</td>
</tr>
<tr class="even">
<td>sample</td>
<td>The name of the sample the gene expression was measured in</td>
</tr>
<tr class="odd">
<td>expression</td>
<td>The value of the gene expression</td>
</tr>
<tr class="even">
<td>organism</td>
<td>The organism/species - here all data stem from mice</td>
</tr>
<tr class="odd">
<td>age</td>
<td>The age of the mouse (all mice were 8 weeks here)</td>
</tr>
<tr class="even">
<td>sex</td>
<td>The sex of the mouse</td>
</tr>
<tr class="odd">
<td>infection</td>
<td>The infection state of the mouse, i.e. infected with Influenza A or not infected.</td>
</tr>
<tr class="even">
<td>strain</td>
<td>The Influenza A strain; C57BL/6 in all cases.</td>
</tr>
<tr class="odd">
<td>time</td>
<td>The duration of the infection (in days).</td>
</tr>
<tr class="even">
<td>tissue</td>
<td>The tissue that was used for the gene expression experiment, i.e. cerebellum or spinal cord.</td>
</tr>
<tr class="odd">
<td>mouse</td>
<td>The mouse unique identifier.</td>
</tr>
<tr class="even">
<td>ENTREZID</td>
<td>The gene ID for the ENTREZ database</td>
</tr>
<tr class="odd">
<td>product</td>
<td>The gene product</td>
</tr>
<tr class="even">
<td>ensembl_gene_id</td>
<td>The ID of the gene from the ENSEMBL database</td>
</tr>
<tr class="odd">
<td>external_synonym</td>
<td>A name synonym for the gene</td>
</tr>
<tr class="even">
<td>chromosome_name</td>
<td>The chromosome name of the gene</td>
</tr>
<tr class="odd">
<td>gene_biotype</td>
<td>The gene biotype</td>
</tr>
<tr class="even">
<td>phenotype_description</td>
<td>The phenotype description of the gene</td>
</tr>
<tr class="odd">
<td>hsapiens_homolog_associated_gene_name</td>
<td>The human homologous gene</td>
</tr>
</tbody>
</table>
<p>We are going to use the R function <code>download.file()</code> to download the
CSV file that contains the gene expression data, and we will use
<code>read.csv()</code> to load into memory the content of the CSV file as an
object of class <code>data.frame</code>. Inside the <code>download.file</code> command, the
first entry is a character string with the source URL. This source URL
downloads a CSV file from a GitHub repository. The text after the
comma (<code>"data/rnaseq.csv"</code>) is the destination of the file on your
local machine. You’ll need to have a folder on your machine called
<code>"data"</code> where you’ll download the file. So this command downloads the
remote file, names it <code>"rnaseq.csv"</code> and adds it to a preexisting
folder named <code>"data"</code>.</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="sec-startdata.html#cb206-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">"data/rnaseq.csv"</span>))</span>
<span id="cb206-2"><a href="sec-startdata.html#cb206-2" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/master/data/GSE96870/rnaseq.csv"</span>,</span>
<span id="cb206-3"><a href="sec-startdata.html#cb206-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"data/rnaseq.csv"</span>)</span></code></pre></div>
<p>You are now ready to load the data:</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="sec-startdata.html#cb207-1" aria-hidden="true" tabindex="-1"></a>rna <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"data/rnaseq.csv"</span>)</span></code></pre></div>
<p>This statement doesn’t produce any output because, as you might
recall, assignments don’t display anything. If we want to check that
our data has been loaded, we can see the contents of the data frame by
typing its name:</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="sec-startdata.html#cb208-1" aria-hidden="true" tabindex="-1"></a>rna</span></code></pre></div>
<p>Wow… that was a lot of output. At least it means the data loaded
properly. Let’s check the top (the first 6 lines) of this data frame
using the function <code>head()</code>:</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="sec-startdata.html#cb209-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(rna)</span></code></pre></div>
<pre><code>##      gene     sample expression     organism age    sex  infection  strain time
## 1     Asl GSM2545336       1170 Mus musculus   8 Female InfluenzaA C57BL/6    8
## 2    Apod GSM2545336      36194 Mus musculus   8 Female InfluenzaA C57BL/6    8
## 3 Cyp2d22 GSM2545336       4060 Mus musculus   8 Female InfluenzaA C57BL/6    8
## 4    Klk6 GSM2545336        287 Mus musculus   8 Female InfluenzaA C57BL/6    8
## 5   Fcrls GSM2545336         85 Mus musculus   8 Female InfluenzaA C57BL/6    8
## 6  Slc2a4 GSM2545336        782 Mus musculus   8 Female InfluenzaA C57BL/6    8
##       tissue mouse ENTREZID
## 1 Cerebellum    14   109900
## 2 Cerebellum    14    11815
## 3 Cerebellum    14    56448
## 4 Cerebellum    14    19144
## 5 Cerebellum    14    80891
## 6 Cerebellum    14    20528
##                                                                    product
## 1                            argininosuccinate lyase transcript variant X1
## 2                                    apolipoprotein D transcript variant 3
## 3 cytochrome P450 family 2 subfamily d polypeptide 22 transcript variant 2
## 4                      kallikrein related-peptidase 6 transcript variant 2
## 5              Fc receptor-like S scavenger receptor transcript variant X1
## 6       solute carrier family 2 (facilitated glucose transporter) member 4
##      ensembl_gene_id external_synonym chromosome_name   gene_biotype
## 1 ENSMUSG00000025533    2510006M18Rik               5 protein_coding
## 2 ENSMUSG00000022548             &lt;NA&gt;              16 protein_coding
## 3 ENSMUSG00000061740             2D22              15 protein_coding
## 4 ENSMUSG00000050063             Bssp               7 protein_coding
## 5 ENSMUSG00000015852    2810439C17Rik               3 protein_coding
## 6 ENSMUSG00000018566           Glut-4              11 protein_coding
##                             phenotype_description
## 1           abnormal circulating amino acid level
## 2                      abnormal lipid homeostasis
## 3                        abnormal skin morphology
## 4                         abnormal cytokine level
## 5 decreased CD8-positive alpha-beta T cell number
## 6              abnormal circulating glucose level
##   hsapiens_homolog_associated_gene_name
## 1                                   ASL
## 2                                  APOD
## 3                                CYP2D6
## 4                                  KLK6
## 5                                 FCRL4
## 6                                SLC2A4</code></pre>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="sec-startdata.html#cb211-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Try also</span></span>
<span id="cb211-2"><a href="sec-startdata.html#cb211-2" aria-hidden="true" tabindex="-1"></a><span class="do">## View(rna)</span></span></code></pre></div>
<p><strong>Note</strong></p>
<p><code>read.csv()</code> assumes that fields are delineated by commas, however, in
several countries, the comma is used as a decimal separator and the
semicolon (;) is used as a field delineator. If you want to read in
this type of files in R, you can use the <code>read.csv2()</code> function. It
behaves exactly like <code>read.csv()</code> but uses different parameters for
the decimal and the field separators. If you are working with another
format, they can be both specified by the user. Check out the help for
<code>read.csv()</code> by typing <code>?read.csv</code> to learn more. There is also the
<code>read.delim()</code> for in tab separated data files. It is important to
note that all of these functions are actually wrapper functions for
the main <code>read.table()</code> function with different arguments. As such,
the data above could have also been loaded by using <code>read.table()</code>
with the separation argument as <code>,</code>. The code is as follows:</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="sec-startdata.html#cb212-1" aria-hidden="true" tabindex="-1"></a>rna <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">"data/rnaseq.csv"</span>,</span>
<span id="cb212-2"><a href="sec-startdata.html#cb212-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">sep =</span> <span class="st">","</span>, <span class="at">quote =</span> <span class="st">"</span><span class="sc">\"</span><span class="st">"</span>,</span>
<span id="cb212-3"><a href="sec-startdata.html#cb212-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">header =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>The header argument has to be set to <code>TRUE</code> to be able to read the
headers as by default <code>read.table()</code> has the header argument set to
FALSE. The quote argument has to be set to <code>"\""</code> to only allow "
as a quoting character (see how is written the <code>product</code> variable
for gene <code>Rtca</code> in rnaseq.csv).</p>
</div>
<div id="what-are-data-frames" class="section level2" number="4.2">
<h2>
<span class="header-section-number">4.2</span> What are data frames?<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('what-are-data-frames')" onmouseout="reset_tooltip('what-are-data-frames-tooltip')"><span class="tooltiptext" id="what-are-data-frames-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Data frames are the <em>de facto</em> data structure for most tabular data,
and what we use for statistics and plotting.</p>
<p>A data frame can be created by hand, but most commonly they are
generated by the functions <code>read.csv()</code> or <code>read.table()</code>; in other
words, when importing spreadsheets from your hard drive (or the web).</p>
<p>A data frame is the representation of data in the format of a table
where the columns are vectors that all have the same length. Because
columns are vectors, each column must contain a single type of data
(e.g., characters, integers, factors). For example, here is a figure
depicting a data frame comprising a numeric, a character, and a
logical vector.</p>
<p><img src="figs/data-frame.svg"></p>
<p>We can see this when inspecting the <b>str</b>ucture of a data frame
with the function <code>str()</code>:</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="sec-startdata.html#cb213-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(rna)</span></code></pre></div>
<pre><code>## 'data.frame':    32428 obs. of  19 variables:
##  $ gene                                 : chr  "Asl" "Apod" "Cyp2d22" "Klk6" ...
##  $ sample                               : chr  "GSM2545336" "GSM2545336" "GSM2545336" "GSM2545336" ...
##  $ expression                           : int  1170 36194 4060 287 85 782 1619 288 43217 1071 ...
##  $ organism                             : chr  "Mus musculus" "Mus musculus" "Mus musculus" "Mus musculus" ...
##  $ age                                  : int  8 8 8 8 8 8 8 8 8 8 ...
##  $ sex                                  : chr  "Female" "Female" "Female" "Female" ...
##  $ infection                            : chr  "InfluenzaA" "InfluenzaA" "InfluenzaA" "InfluenzaA" ...
##  $ strain                               : chr  "C57BL/6" "C57BL/6" "C57BL/6" "C57BL/6" ...
##  $ time                                 : int  8 8 8 8 8 8 8 8 8 8 ...
##  $ tissue                               : chr  "Cerebellum" "Cerebellum" "Cerebellum" "Cerebellum" ...
##  $ mouse                                : int  14 14 14 14 14 14 14 14 14 14 ...
##  $ ENTREZID                             : int  109900 11815 56448 19144 80891 20528 97827 118454 18823 14696 ...
##  $ product                              : chr  "argininosuccinate lyase transcript variant X1" "apolipoprotein D transcript variant 3" "cytochrome P450 family 2 subfamily d polypeptide 22 transcript variant 2" "kallikrein related-peptidase 6 transcript variant 2" ...
##  $ ensembl_gene_id                      : chr  "ENSMUSG00000025533" "ENSMUSG00000022548" "ENSMUSG00000061740" "ENSMUSG00000050063" ...
##  $ external_synonym                     : chr  "2510006M18Rik" NA "2D22" "Bssp" ...
##  $ chromosome_name                      : chr  "5" "16" "15" "7" ...
##  $ gene_biotype                         : chr  "protein_coding" "protein_coding" "protein_coding" "protein_coding" ...
##  $ phenotype_description                : chr  "abnormal circulating amino acid level" "abnormal lipid homeostasis" "abnormal skin morphology" "abnormal cytokine level" ...
##  $ hsapiens_homolog_associated_gene_name: chr  "ASL" "APOD" "CYP2D6" "KLK6" ...</code></pre>
</div>
<div id="inspecting-data.frame-objects" class="section level2" number="4.3">
<h2>
<span class="header-section-number">4.3</span> Inspecting <code>data.frame</code> Objects<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('inspecting-data.frame-objects')" onmouseout="reset_tooltip('inspecting-data.frame-objects-tooltip')"><span class="tooltiptext" id="inspecting-data.frame-objects-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>We already saw how the functions <code>head()</code> and <code>str()</code> can be useful to
check the content and the structure of a data frame. Here is a
non-exhaustive list of functions to get a sense of the
content/structure of the data. Let’s try them out!</p>
<p><strong>Size</strong>:</p>
<ul>
<li>
<code>dim(rna)</code> - returns a vector with the number of rows in the first
element, and the number of columns as the second element (the
<strong>dim</strong>ensions of the object)</li>
<li>
<code>nrow(rna)</code> - returns the number of rows</li>
<li>
<code>ncol(rna)</code> - returns the number of columns</li>
</ul>
<p><strong>Content</strong>:</p>
<ul>
<li>
<code>head(rna)</code> - shows the first 6 rows</li>
<li>
<code>tail(rna)</code> - shows the last 6 rows</li>
</ul>
<p><strong>Names</strong>:</p>
<ul>
<li>
<code>names(rna)</code> - returns the column names (synonym of <code>colnames()</code> for
<code>data.frame</code> objects)</li>
<li>
<code>rownames(rna)</code> - returns the row names</li>
</ul>
<p><strong>Summary</strong>:</p>
<ul>
<li>
<code>str(rna)</code> - structure of the object and information about the
class, length and content of each column</li>
<li>
<code>summary(rna)</code> - summary statistics for each column</li>
</ul>
<p>Note: most of these functions are “generic,” they can be used on other types of
objects besides <code>data.frame</code>.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Based on the output of <code>str(rna)</code>, can you answer the following
questions?</p>
<ul>
<li>What is the class of the object <code>rna</code>?</li>
<li>How many rows and how many columns are in this object?</li>
<li>How many genes (as defined by the <code>gene</code> variable) have been
measured in this experiment?</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-16" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-16', 'sol-start-16')"></span>
</p>
<div id="sol-body-16" class="solution-body" style="display: none;">
<ul>
<li>class: data frame</li>
<li>how many rows: 32428, how many columns: 21</li>
<li>how many genes: 1474</li>
</ul>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="indexing-and-subsetting-data-frames" class="section level2" number="4.4">
<h2>
<span class="header-section-number">4.4</span> Indexing and subsetting data frames<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('indexing-and-subsetting-data-frames')" onmouseout="reset_tooltip('indexing-and-subsetting-data-frames-tooltip')"><span class="tooltiptext" id="indexing-and-subsetting-data-frames-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Our <code>rna</code> data frame has rows and columns (it has 2 dimensions), if we
want to extract some specific data from it, we need to specify the
“coordinates” we want from it. Row numbers come first, followed by
column numbers. However, note that different ways of specifying these
coordinates lead to results with different classes.</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="sec-startdata.html#cb215-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first element in the first column of the data frame (as a vector)</span></span>
<span id="cb215-2"><a href="sec-startdata.html#cb215-2" aria-hidden="true" tabindex="-1"></a>rna[<span class="dv">1</span>, <span class="dv">1</span>]</span>
<span id="cb215-3"><a href="sec-startdata.html#cb215-3" aria-hidden="true" tabindex="-1"></a><span class="co"># first element in the 6th column (as a vector)</span></span>
<span id="cb215-4"><a href="sec-startdata.html#cb215-4" aria-hidden="true" tabindex="-1"></a>rna[<span class="dv">1</span>, <span class="dv">6</span>]</span>
<span id="cb215-5"><a href="sec-startdata.html#cb215-5" aria-hidden="true" tabindex="-1"></a><span class="co"># first column of the data frame (as a vector)</span></span>
<span id="cb215-6"><a href="sec-startdata.html#cb215-6" aria-hidden="true" tabindex="-1"></a>rna[, <span class="dv">1</span>]</span>
<span id="cb215-7"><a href="sec-startdata.html#cb215-7" aria-hidden="true" tabindex="-1"></a><span class="co"># first column of the data frame (as a data.frame)</span></span>
<span id="cb215-8"><a href="sec-startdata.html#cb215-8" aria-hidden="true" tabindex="-1"></a>rna[<span class="dv">1</span>]</span>
<span id="cb215-9"><a href="sec-startdata.html#cb215-9" aria-hidden="true" tabindex="-1"></a><span class="co"># first three elements in the 7th column (as a vector)</span></span>
<span id="cb215-10"><a href="sec-startdata.html#cb215-10" aria-hidden="true" tabindex="-1"></a>rna[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">7</span>]</span>
<span id="cb215-11"><a href="sec-startdata.html#cb215-11" aria-hidden="true" tabindex="-1"></a><span class="co"># the 3rd row of the data frame (as a data.frame)</span></span>
<span id="cb215-12"><a href="sec-startdata.html#cb215-12" aria-hidden="true" tabindex="-1"></a>rna[<span class="dv">3</span>, ]</span>
<span id="cb215-13"><a href="sec-startdata.html#cb215-13" aria-hidden="true" tabindex="-1"></a><span class="co"># equivalent to head_rna &lt;- head(rna)</span></span>
<span id="cb215-14"><a href="sec-startdata.html#cb215-14" aria-hidden="true" tabindex="-1"></a>head_rna <span class="ot">&lt;-</span> rna[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, ]</span>
<span id="cb215-15"><a href="sec-startdata.html#cb215-15" aria-hidden="true" tabindex="-1"></a>head_rna</span></code></pre></div>
<p><code>:</code> is a special function that creates numeric vectors of integers in
increasing or decreasing order, test <code>1:10</code> and <code>10:1</code> for
instance. See section <a href="sec-startr.html#sec-genvec">3.9</a> for details.</p>
<p>You can also exclude certain indices of a data frame using the “<code>-</code>” sign:</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="sec-startdata.html#cb216-1" aria-hidden="true" tabindex="-1"></a>rna[, <span class="sc">-</span><span class="dv">1</span>]          <span class="do">## The whole data frame, except the first column</span></span>
<span id="cb216-2"><a href="sec-startdata.html#cb216-2" aria-hidden="true" tabindex="-1"></a>rna[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">7</span><span class="sc">:</span><span class="dv">32428</span>), ] <span class="do">## Equivalent to head(rna)</span></span></code></pre></div>
<p>Data frames can be subset by calling indices (as shown previously),
but also by calling their column names directly:</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="sec-startdata.html#cb217-1" aria-hidden="true" tabindex="-1"></a>rna[<span class="st">"gene"</span>]       <span class="co"># Result is a data.frame</span></span>
<span id="cb217-2"><a href="sec-startdata.html#cb217-2" aria-hidden="true" tabindex="-1"></a>rna[, <span class="st">"gene"</span>]     <span class="co"># Result is a vector</span></span>
<span id="cb217-3"><a href="sec-startdata.html#cb217-3" aria-hidden="true" tabindex="-1"></a>rna[[<span class="st">"gene"</span>]]     <span class="co"># Result is a vector</span></span>
<span id="cb217-4"><a href="sec-startdata.html#cb217-4" aria-hidden="true" tabindex="-1"></a>rna<span class="sc">$</span>gene          <span class="co"># Result is a vector</span></span></code></pre></div>
<p>In RStudio, you can use the autocompletion feature to get the full and
correct names of the columns.</p>
<p>When we inspect the elements of the column
<code>hsapiens_homolog_associated_gene_name</code> (for example with <code>View(rna)</code>),
we can see that some cells contain NA values. If we wanted to extract
only mouse genes of this table that have a human homologous,
we could combine <code>is.na()</code> and data frames subsetting:</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="sec-startdata.html#cb218-1" aria-hidden="true" tabindex="-1"></a>is_missing_hsapiens_homolog <span class="ot">&lt;-</span> <span class="fu">is.na</span>(rna<span class="sc">$</span>hsapiens_homolog_associated_gene_name)</span>
<span id="cb218-2"><a href="sec-startdata.html#cb218-2" aria-hidden="true" tabindex="-1"></a>rna_hsapiens_homolog <span class="ot">&lt;-</span> rna[<span class="sc">!</span>is_missing_hsapiens_homolog,]</span></code></pre></div>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="sec-startdata.html#cb219-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(rna_hsapiens_homolog)</span></code></pre></div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>How many mouse genes do not have a human homologous?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-17" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-17', 'sol-start-17')"></span>
</p>
<div id="sol-body-17" class="solution-body" style="display: none;">
<blockquote>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="sec-startdata.html#cb220-1" aria-hidden="true" tabindex="-1"></a>rna_missing_hsapiens_homolog <span class="ot">&lt;-</span> rna[is_missing_hsapiens_homolog,]</span>
<span id="cb220-2"><a href="sec-startdata.html#cb220-2" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(rna_missing_hsapiens_homolog)</span></code></pre></div>
<pre><code>## [1] 4774</code></pre>
</blockquote>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ol style="list-style-type: decimal">
<li><p>Create a <code>data.frame</code> (<code>rna_200</code>) containing only the data in
row 200 of the <code>rna</code> dataset.</p></li>
<li><p>Notice how <code>nrow()</code> gave you the number of rows in a <code>data.frame</code>?</p></li>
</ol>
<ul>
<li><p>Use that number to pull out just that last row in the initial
<code>rna</code> data frame.</p></li>
<li><p>Compare that with what you see as the last row using <code>tail()</code> to
make sure it’s meeting expectations.</p></li>
<li><p>Pull out that last row using <code>nrow()</code> instead of the row number.</p></li>
<li><p>Create a new data frame (<code>rna_last</code>) from that last row.</p></li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li><p>Use <code>nrow()</code> to extract the row that is in the middle of the
<code>rna</code> dataframe. Store the content of this row in an object
named <code>rna_middle</code>.</p></li>
<li><p>Combine <code>nrow()</code> with the <code>-</code> notation above to reproduce the
behavior of <code>head(rna)</code>, keeping just the first through 6th
rows of the rna dataset.</p></li>
</ol>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-18" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-18', 'sol-start-18')"></span>
</p>
<div id="sol-body-18" class="solution-body" style="display: none;">
<blockquote>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="sec-startdata.html#cb222-1" aria-hidden="true" tabindex="-1"></a><span class="do">## 1.</span></span>
<span id="cb222-2"><a href="sec-startdata.html#cb222-2" aria-hidden="true" tabindex="-1"></a>rna_200 <span class="ot">&lt;-</span> rna[<span class="dv">200</span>, ]</span>
<span id="cb222-3"><a href="sec-startdata.html#cb222-3" aria-hidden="true" tabindex="-1"></a><span class="do">## 2.</span></span>
<span id="cb222-4"><a href="sec-startdata.html#cb222-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Saving `n_rows` to improve readability and reduce duplication</span></span>
<span id="cb222-5"><a href="sec-startdata.html#cb222-5" aria-hidden="true" tabindex="-1"></a>n_rows <span class="ot">&lt;-</span> <span class="fu">nrow</span>(rna)</span>
<span id="cb222-6"><a href="sec-startdata.html#cb222-6" aria-hidden="true" tabindex="-1"></a>rna_last <span class="ot">&lt;-</span> rna[n_rows, ]</span>
<span id="cb222-7"><a href="sec-startdata.html#cb222-7" aria-hidden="true" tabindex="-1"></a><span class="do">## 3.</span></span>
<span id="cb222-8"><a href="sec-startdata.html#cb222-8" aria-hidden="true" tabindex="-1"></a>rna_middle <span class="ot">&lt;-</span> rna[n_rows <span class="sc">/</span> <span class="dv">2</span>, ]</span>
<span id="cb222-9"><a href="sec-startdata.html#cb222-9" aria-hidden="true" tabindex="-1"></a><span class="do">## 4.</span></span>
<span id="cb222-10"><a href="sec-startdata.html#cb222-10" aria-hidden="true" tabindex="-1"></a>rna_head <span class="ot">&lt;-</span> rna[<span class="sc">-</span>(<span class="dv">7</span><span class="sc">:</span>n_rows), ]</span></code></pre></div>
</blockquote>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="factors" class="section level2" number="4.5">
<h2>
<span class="header-section-number">4.5</span> Factors<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('factors')" onmouseout="reset_tooltip('factors-tooltip')"><span class="tooltiptext" id="factors-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Factors represent <strong>categorical data</strong>. They are stored as integers
associated with labels and they can be ordered or unordered. While
factors look (and often behave) like character vectors, they are
actually treated as integer vectors by R. So you need to be very
careful when treating them as strings.</p>
<p>Once created, factors can only contain a pre-defined set of values,
known as <em>levels</em>. By default, R always sorts levels in alphabetical
order. For instance, if you have a factor with 2 levels:</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="sec-startdata.html#cb223-1" aria-hidden="true" tabindex="-1"></a>sex <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"male"</span>, <span class="st">"female"</span>, <span class="st">"female"</span>, <span class="st">"male"</span>, <span class="st">"female"</span>))</span></code></pre></div>
<p>The levels can be seen this by using the function <code>levels()</code> and you can find the number of levels using <code>nlevels()</code>:</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="sec-startdata.html#cb224-1" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(sex)</span></code></pre></div>
<pre><code>## [1] "female" "male"</code></pre>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="sec-startdata.html#cb226-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nlevels</span>(sex)</span></code></pre></div>
<pre><code>## [1] 2</code></pre>
<p>Sometimes, the order of the factors does not matter, other times you
might want to specify the order because it is meaningful (e.g., “low,”
“medium,” “high”), it improves your visualization, or it is required
by a particular type of analysis. Here, one way to reorder our levels
in the <code>sex</code> vector would be:</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="sec-startdata.html#cb228-1" aria-hidden="true" tabindex="-1"></a>sex <span class="do">## current order</span></span></code></pre></div>
<pre><code>## [1] male   female female male   female
## Levels: female male</code></pre>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="sec-startdata.html#cb230-1" aria-hidden="true" tabindex="-1"></a>sex <span class="ot">&lt;-</span> <span class="fu">factor</span>(sex, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"male"</span>, <span class="st">"female"</span>))</span>
<span id="cb230-2"><a href="sec-startdata.html#cb230-2" aria-hidden="true" tabindex="-1"></a>sex <span class="do">## after re-ordering</span></span></code></pre></div>
<pre><code>## [1] male   female female male   female
## Levels: male female</code></pre>
<div id="converting-to-factors" class="section level3 unnumbered">
<h3>Converting to factors<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('converting-to-factors')" onmouseout="reset_tooltip('converting-to-factors-tooltip')"><span class="tooltiptext" id="converting-to-factors-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>If you need to convert a factor to a character vector, you use
<code>as.character(x)</code>.</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="sec-startdata.html#cb232-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.character</span>(sex)</span></code></pre></div>
<pre><code>## [1] "male"   "female" "female" "male"   "female"</code></pre>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="sec-startdata.html#cb234-1" aria-hidden="true" tabindex="-1"></a><span class="co"># try also: as.numeric(sex)</span></span></code></pre></div>
<!-- ### Numeric factors {-} -->
<!-- Converting factors where the levels appear as numbers (such as -->
<!-- concentration levels, or years) to a numeric vector is a little -->
<!-- trickier. The `as.numeric()` function returns the index values of the -->
<!-- factor, not its levels, so it will result in an entirely new (and -->
<!-- unwanted in this case) set of numbers.  One method to avoid this is to -->
<!-- convert factors to characters, and then to numbers.  Another method is -->
<!-- to use the `levels()` function. Compare: -->
<!-- ```{r} -->
<!-- year_fct <- factor(c(1990, 1983, 1977, 1998, 1990)) -->
<!-- as.numeric(year_fct)  ## Wrong! And there is no warning... -->
<!-- as.numeric(as.character(year_fct)) ## Works... -->
<!-- as.numeric(levels(year_fct))[year_fct] ## The recommended way. -->
<!-- ```

<!-- Notice that in the `levels()` approach, three important steps occur: -->
<!-- * We obtain all the factor levels using `levels(year_fct)` -->
<!-- * We convert these levels to numeric values using `as.numeric(levels(year_fct))` -->
<!-- * We then access these numeric values using the underlying integers of the -->
<!--   vector `year_fct` inside the square brackets -->
</div>
<div id="renaming-factors" class="section level3 unnumbered">
<h3>Renaming factors<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('renaming-factors')" onmouseout="reset_tooltip('renaming-factors-tooltip')"><span class="tooltiptext" id="renaming-factors-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>When your data is stored as a factor, you can use the <code>plot()</code>
function to get a quick glance at the number of observations
represented by each factor level. Let’s look at the number of males
and females in our data.</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="sec-startdata.html#cb235-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sex)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:firstfactorplot"></span>
<p class="caption marginnote shownote">
Figure 4.1: Bar plot of the number of females and males.
</p>
<img src="BIOINFO_01_files/figure-html/firstfactorplot-1.png" alt="Bar plot of the number of females and males." width="672">
</div>
<p>If we want to rename these factor, it is sufficient to change its
levels:</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="sec-startdata.html#cb236-1" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(sex)</span></code></pre></div>
<pre><code>## [1] "male"   "female"</code></pre>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="sec-startdata.html#cb238-1" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(sex) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"F"</span>)</span>
<span id="cb238-2"><a href="sec-startdata.html#cb238-2" aria-hidden="true" tabindex="-1"></a>sex</span></code></pre></div>
<pre><code>## [1] M F F M F
## Levels: M F</code></pre>
<div class="sourceCode" id="cb240"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb240-1"><a href="sec-startdata.html#cb240-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sex)</span></code></pre></div>
<p><img src="BIOINFO_01_files/figure-html/unnamed-chunk-100-1.png" width="672"></p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li>Rename “female” and “male” to “Female” and “Male” respectively.</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>We have seen how data frames are created when using <code>read.csv()</code>, but
they can also be created by hand with the <code>data.frame()</code> function.
There are a few mistakes in this hand-crafted <code>data.frame</code>. Can you
spot and fix them? Don’t hesitate to experiment!</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="sec-startdata.html#cb241-1" aria-hidden="true" tabindex="-1"></a>animal_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb241-2"><a href="sec-startdata.html#cb241-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">animal =</span> <span class="fu">c</span>(dog, cat, sea cucumber, sea urchin),</span>
<span id="cb241-3"><a href="sec-startdata.html#cb241-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">feel =</span> <span class="fu">c</span>(<span class="st">"furry"</span>, <span class="st">"squishy"</span>, <span class="st">"spiny"</span>),</span>
<span id="cb241-4"><a href="sec-startdata.html#cb241-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">weight =</span> <span class="fu">c</span>(<span class="dv">45</span>, <span class="dv">8</span> <span class="fl">1.1</span>, <span class="fl">0.8</span>))</span></code></pre></div>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-19" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-19', 'sol-start-19')"></span>
</p>
<div id="sol-body-19" class="solution-body" style="display: none;">
<ul>
<li>missing quotations around the names of the animals</li>
<li>missing one entry in the “feel” column (probably for one of the furry animals)</li>
<li>missing one comma in the weight column</li>
</ul>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Can you predict the class for each of the columns in the following
example?</p>
<p>Check your guesses using <code>str(country_climate)</code>:</p>
<ul>
<li><p>Are they what you expected? Why? Why not?</p></li>
<li><p>Try again by adding <code>stringsAsFactors = TRUE</code> after the last
variable when creating the data frame? What is happening now?
<code>stringsAsFactors</code> can also be set when reading text-based
spreadsheets into R using <code>read.csv()</code>.</p></li>
</ul>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="sec-startdata.html#cb242-1" aria-hidden="true" tabindex="-1"></a>country_climate <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb242-2"><a href="sec-startdata.html#cb242-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">country =</span> <span class="fu">c</span>(<span class="st">"Canada"</span>, <span class="st">"Panama"</span>, <span class="st">"South Africa"</span>, <span class="st">"Australia"</span>),</span>
<span id="cb242-3"><a href="sec-startdata.html#cb242-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">climate =</span> <span class="fu">c</span>(<span class="st">"cold"</span>, <span class="st">"hot"</span>, <span class="st">"temperate"</span>, <span class="st">"hot/temperate"</span>),</span>
<span id="cb242-4"><a href="sec-startdata.html#cb242-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">temperature =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">30</span>, <span class="dv">18</span>, <span class="st">"15"</span>),</span>
<span id="cb242-5"><a href="sec-startdata.html#cb242-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">northern_hemisphere =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="st">"FALSE"</span>),</span>
<span id="cb242-6"><a href="sec-startdata.html#cb242-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">has_kangaroo =</span> <span class="fu">c</span>(<span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="dv">1</span>)</span>
<span id="cb242-7"><a href="sec-startdata.html#cb242-7" aria-hidden="true" tabindex="-1"></a>       )</span></code></pre></div>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>The automatic conversion of data type is sometimes a blessing, sometimes an
annoyance. Be aware that it exists, learn the rules, and double check that data
you import in R are of the correct type within your data frame. If not, use it
to your advantage to detect mistakes that might have been introduced during data
entry (a letter in a column that should only contain numbers for instance).</p>
<p>Learn more in this <a href="https://support.rstudio.com/hc/en-us/articles/218611977-Importing-Data-with-RStudio">RStudio
tutorial</a></p>
</div>
</div>
<div id="matrices" class="section level2" number="4.6">
<h2>
<span class="header-section-number">4.6</span> Matrices<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('matrices')" onmouseout="reset_tooltip('matrices-tooltip')"><span class="tooltiptext" id="matrices-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Before proceeding, now that we have learnt about dataframes, let’s
recap package installation and learn about a new data type, namely the
<code>matrix</code>. Like a <code>data.frame</code>, a matrix has two dimensions, rows and
columns. But the major difference is that all cells in a <code>matrix</code> must
be of the same type: <code>numeric</code>, <code>character</code>, <code>logical</code>, … In that
respect, matrices are closer to a <code>vector</code> than a <code>data.frame</code>.</p>
<p>The default constructor for a matrix is <code>matrix</code>. It takes a vector of
values to populate the matrix and the number of row and/or
columns<label for="tufte-sn-7" class="margin-toggle sidenote-number">7</label><input type="checkbox" id="tufte-sn-7" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">7</span> Either the number of rows or columns are enough, as the other
one can be deduced from the length of the values. Try out what happens
if the values and number of rows/columns don’t add up.</span>. The values are sorted along the columns, as illustrated
below but you can also sort them along the row with the argument <code>byrow = TRUE</code>.</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="sec-startdata.html#cb243-1" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>, <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">nrow =</span> <span class="dv">3</span>)</span>
<span id="cb243-2"><a href="sec-startdata.html#cb243-2" aria-hidden="true" tabindex="-1"></a>m</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3]
## [1,]    1    4    7
## [2,]    2    5    8
## [3,]    3    6    9</code></pre>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="sec-startdata.html#cb245-1" aria-hidden="true" tabindex="-1"></a><span class="co"># try now with byrow = TRUE</span></span></code></pre></div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>
Using the function <code>installed.packages()</code>, create a <code>character</code> matrix containing the information about all packages currently installed on your computer. Explore it.
</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-20" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-20', 'sol-start-20')"></span>
</p>
<div id="sol-body-20" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="sec-startdata.html#cb246-1" aria-hidden="true" tabindex="-1"></a><span class="do">## create the matrix</span></span>
<span id="cb246-2"><a href="sec-startdata.html#cb246-2" aria-hidden="true" tabindex="-1"></a>ip <span class="ot">&lt;-</span> <span class="fu">installed.packages</span>()</span>
<span id="cb246-3"><a href="sec-startdata.html#cb246-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ip)</span>
<span id="cb246-4"><a href="sec-startdata.html#cb246-4" aria-hidden="true" tabindex="-1"></a><span class="do">## try also View(ip)</span></span>
<span id="cb246-5"><a href="sec-startdata.html#cb246-5" aria-hidden="true" tabindex="-1"></a><span class="do">## number of package</span></span>
<span id="cb246-6"><a href="sec-startdata.html#cb246-6" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(ip)</span>
<span id="cb246-7"><a href="sec-startdata.html#cb246-7" aria-hidden="true" tabindex="-1"></a><span class="do">## names of all installed packages</span></span>
<span id="cb246-8"><a href="sec-startdata.html#cb246-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ip)</span>
<span id="cb246-9"><a href="sec-startdata.html#cb246-9" aria-hidden="true" tabindex="-1"></a><span class="do">## type of information we have about each package</span></span>
<span id="cb246-10"><a href="sec-startdata.html#cb246-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ip)</span></code></pre></div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>It is often useful to create large random data matrices as test
data. The exercise below asks you to create such a matrix with random
data drawn from a normal distribution of mean 0 and standard deviation
1, which can be done with the <code>rnorm()</code> function.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>
Construct a matrix of dimension 1000 by 3 of normally distributed data (mean 0, standard deviation 1).
</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-21" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-21', 'sol-start-21')"></span>
</p>
<div id="sol-body-21" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="sec-startdata.html#cb247-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb247-2"><a href="sec-startdata.html#cb247-2" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(<span class="dv">3000</span>), <span class="at">ncol =</span> <span class="dv">3</span>)</span>
<span id="cb247-3"><a href="sec-startdata.html#cb247-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(m)</span></code></pre></div>
<pre><code>## [1] 1000    3</code></pre>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb249-1"><a href="sec-startdata.html#cb249-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(m)</span></code></pre></div>
<pre><code>##             [,1]        [,2]       [,3]
## [1,] -0.56047565 -0.99579872 -0.5116037
## [2,] -0.23017749 -1.03995504  0.2369379
## [3,]  1.55870831 -0.01798024 -0.5415892
## [4,]  0.07050839 -0.13217513  1.2192276
## [5,]  0.12928774 -2.54934277  0.1741359
## [6,]  1.71506499  1.04057346 -0.6152683</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="summary-of-r-objects" class="section level2" number="4.7">
<h2>
<span class="header-section-number">4.7</span> Summary of R objects<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('summary-of-r-objects')" onmouseout="reset_tooltip('summary-of-r-objects-tooltip')"><span class="tooltiptext" id="summary-of-r-objects-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>So far, we have seen several types of R object varying in the number
of dimensions and whether they could store a single or multiple data
types:</p>
<ul>
<li>
<strong><code>vector</code></strong>: one dimension (they have a length), single type of data.</li>
<li>
<strong><code>matrix</code></strong>: two dimensions, single type of data.</li>
<li>
<strong><code>data.frame</code></strong>: two dimensions, one type per column.</li>
</ul>
</div>
<div id="lists" class="section level2" number="4.8">
<h2>
<span class="header-section-number">4.8</span> Lists<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('lists')" onmouseout="reset_tooltip('lists-tooltip')"><span class="tooltiptext" id="lists-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>A data type that we haven’t seen yet, but that is useful to know, and
follows from the summary that we have just seen are lists:</p>
<ul>
<li>
<strong><code>list</code></strong>: one dimension, every item can be of a different data
type.</li>
</ul>
<p>Below, let’s create a list containing a vector of numbers, characters,
a matrix, a dataframe and another list:</p>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb251-1"><a href="sec-startdata.html#cb251-1" aria-hidden="true" tabindex="-1"></a>l <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="do">## numeric</span></span>
<span id="cb251-2"><a href="sec-startdata.html#cb251-2" aria-hidden="true" tabindex="-1"></a>          letters, <span class="do">## character</span></span>
<span id="cb251-3"><a href="sec-startdata.html#cb251-3" aria-hidden="true" tabindex="-1"></a>          <span class="fu">installed.packages</span>(), <span class="do">## a matrix</span></span>
<span id="cb251-4"><a href="sec-startdata.html#cb251-4" aria-hidden="true" tabindex="-1"></a>          cars, <span class="do">## a data.frame</span></span>
<span id="cb251-5"><a href="sec-startdata.html#cb251-5" aria-hidden="true" tabindex="-1"></a>          <span class="fu">list</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)) <span class="do">## a list</span></span>
<span id="cb251-6"><a href="sec-startdata.html#cb251-6" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(l)</span></code></pre></div>
<pre><code>## [1] 5</code></pre>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="sec-startdata.html#cb253-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(l)</span></code></pre></div>
<pre><code>## List of 5
##  $ : int [1:10] 1 2 3 4 5 6 7 8 9 10
##  $ : chr [1:26] "a" "b" "c" "d" ...
##  $ : chr [1:574, 1:16] "ALL" "ALS" "AnnotationDbi" "AnnotationFilter" ...
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : chr [1:574] "ALL" "ALS" "AnnotationDbi" "AnnotationFilter" ...
##   .. ..$ : chr [1:16] "Package" "LibPath" "Version" "Priority" ...
##  $ :'data.frame':    50 obs. of  2 variables:
##   ..$ speed: num [1:50] 4 4 7 7 8 9 10 10 10 11 ...
##   ..$ dist : num [1:50] 2 10 4 22 16 10 18 26 34 17 ...
##  $ :List of 3
##   ..$ : num 1
##   ..$ : num 2
##   ..$ : num 3</code></pre>
<p>List subsetting is done using <code>[]</code> to subset a new sub-list or <code>[[]]</code>
to extract a single element of that list (using indices or names, of
the list is named).</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="sec-startdata.html#cb255-1" aria-hidden="true" tabindex="-1"></a>l[[<span class="dv">1</span>]] <span class="do">## first element</span></span></code></pre></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10</code></pre>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="sec-startdata.html#cb257-1" aria-hidden="true" tabindex="-1"></a>l[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="do">## a list of length 2</span></span></code></pre></div>
<pre><code>## [[1]]
##  [1]  1  2  3  4  5  6  7  8  9 10
## 
## [[2]]
##  [1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l" "m" "n" "o" "p" "q" "r" "s"
## [20] "t" "u" "v" "w" "x" "y" "z"</code></pre>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="sec-startdata.html#cb259-1" aria-hidden="true" tabindex="-1"></a>l[<span class="dv">1</span>]   <span class="do">## a list of length 1</span></span></code></pre></div>
<pre><code>## [[1]]
##  [1]  1  2  3  4  5  6  7  8  9 10</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>You can also attribute a name to each element of a list. Give the following names (in that order) to the list elements with the function <code>names()</code>: numbers, alphabet, installed_packages, cars, random. Check if the names were correctly attributed.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-22" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-22', 'sol-start-22')"></span>
</p>
<div id="sol-body-22" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb261"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb261-1"><a href="sec-startdata.html#cb261-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(l) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"numbers"</span>, <span class="st">"alphabet"</span>, <span class="st">"installed_packages"</span>, </span>
<span id="cb261-2"><a href="sec-startdata.html#cb261-2" aria-hidden="true" tabindex="-1"></a>              <span class="st">"cars"</span>, <span class="st">"random"</span>)</span>
<span id="cb261-3"><a href="sec-startdata.html#cb261-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(l)</span></code></pre></div>
<pre><code>## [1] "numbers"            "alphabet"           "installed_packages"
## [4] "cars"               "random"</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Subset the previously defined list <code>l</code>, to keep:</p>
<ul>
<li>a list with all but the matrix of installed packages</li>
<li>only the <code>cars</code> data.frame</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-23" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-23', 'sol-start-23')"></span>
</p>
<div id="sol-body-23" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="sec-startdata.html#cb263-1" aria-hidden="true" tabindex="-1"></a><span class="co"># list with all but the matrix of installed packages</span></span>
<span id="cb263-2"><a href="sec-startdata.html#cb263-2" aria-hidden="true" tabindex="-1"></a>l[<span class="sc">-</span><span class="dv">3</span>]</span></code></pre></div>
<pre><code>## $numbers
##  [1]  1  2  3  4  5  6  7  8  9 10
## 
## $alphabet
##  [1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l" "m" "n" "o" "p" "q" "r" "s"
## [20] "t" "u" "v" "w" "x" "y" "z"
## 
## $cars
##    speed dist
## 1      4    2
## 2      4   10
## 3      7    4
## 4      7   22
## 5      8   16
## 6      9   10
## 7     10   18
## 8     10   26
## 9     10   34
## 10    11   17
## 11    11   28
## 12    12   14
## 13    12   20
## 14    12   24
## 15    12   28
## 16    13   26
## 17    13   34
## 18    13   34
## 19    13   46
## 20    14   26
## 21    14   36
## 22    14   60
## 23    14   80
## 24    15   20
## 25    15   26
## 26    15   54
## 27    16   32
## 28    16   40
## 29    17   32
## 30    17   40
## 31    17   50
## 32    18   42
## 33    18   56
## 34    18   76
## 35    18   84
## 36    19   36
## 37    19   46
## 38    19   68
## 39    20   32
## 40    20   48
## 41    20   52
## 42    20   56
## 43    20   64
## 44    22   66
## 45    23   54
## 46    24   70
## 47    24   92
## 48    24   93
## 49    24  120
## 50    25   85
## 
## $random
## $random[[1]]
## [1] 1
## 
## $random[[2]]
## [1] 2
## 
## $random[[3]]
## [1] 3</code></pre>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="sec-startdata.html#cb265-1" aria-hidden="true" tabindex="-1"></a><span class="co"># only the `cars` data.frame</span></span>
<span id="cb265-2"><a href="sec-startdata.html#cb265-2" aria-hidden="true" tabindex="-1"></a>l[[<span class="st">"cars"</span>]]</span></code></pre></div>
<pre><code>##    speed dist
## 1      4    2
## 2      4   10
## 3      7    4
## 4      7   22
## 5      8   16
## 6      9   10
## 7     10   18
## 8     10   26
## 9     10   34
## 10    11   17
## 11    11   28
## 12    12   14
## 13    12   20
## 14    12   24
## 15    12   28
## 16    13   26
## 17    13   34
## 18    13   34
## 19    13   46
## 20    14   26
## 21    14   36
## 22    14   60
## 23    14   80
## 24    15   20
## 25    15   26
## 26    15   54
## 27    16   32
## 28    16   40
## 29    17   32
## 30    17   40
## 31    17   50
## 32    18   42
## 33    18   56
## 34    18   76
## 35    18   84
## 36    19   36
## 37    19   46
## 38    19   68
## 39    20   32
## 40    20   48
## 41    20   52
## 42    20   56
## 43    20   64
## 44    22   66
## 45    23   54
## 46    24   70
## 47    24   92
## 48    24   93
## 49    24  120
## 50    25   85</code></pre>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="sec-startdata.html#cb267-1" aria-hidden="true" tabindex="-1"></a><span class="co"># equivalent to:</span></span>
<span id="cb267-2"><a href="sec-startdata.html#cb267-2" aria-hidden="true" tabindex="-1"></a>l[[<span class="dv">4</span>]]</span></code></pre></div>
<pre><code>##    speed dist
## 1      4    2
## 2      4   10
## 3      7    4
## 4      7   22
## 5      8   16
## 6      9   10
## 7     10   18
## 8     10   26
## 9     10   34
## 10    11   17
## 11    11   28
## 12    12   14
## 13    12   20
## 14    12   24
## 15    12   28
## 16    13   26
## 17    13   34
## 18    13   34
## 19    13   46
## 20    14   26
## 21    14   36
## 22    14   60
## 23    14   80
## 24    15   20
## 25    15   26
## 26    15   54
## 27    16   32
## 28    16   40
## 29    17   32
## 30    17   40
## 31    17   50
## 32    18   42
## 33    18   56
## 34    18   76
## 35    18   84
## 36    19   36
## 37    19   46
## 38    19   68
## 39    20   32
## 40    20   48
## 41    20   52
## 42    20   56
## 43    20   64
## 44    22   66
## 45    23   54
## 46    24   70
## 47    24   92
## 48    24   93
## 49    24  120
## 50    25   85</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="sec-exportandsave" class="section level2" number="4.9">
<h2>
<span class="header-section-number">4.9</span> Exporting and saving data<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('sec-exportandsave')" onmouseout="reset_tooltip('sec-exportandsave-tooltip')"><span class="tooltiptext" id="sec-exportandsave-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<div id="exporting-tabular-data" class="section level3 unnumbered">
<h3>Exporting tabular data<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('exporting-tabular-data')" onmouseout="reset_tooltip('exporting-tabular-data-tooltip')"><span class="tooltiptext" id="exporting-tabular-data-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>We have seen how to read a text-based spreadsheet into R using the
<code>read.table</code> family of functions. To export a <code>data.frame</code> to a
text-based spreadsheet, we can use the <code>write.table</code> set of functions
(<code>write.csv</code>, <code>write.delim</code>, …). They all take the variable to be
exported and the file to be exported to. For example, to export the
<code>rna</code> data to the <code>my_rnaseq.csv</code> file in the <code>data_output</code>
directory, we would execute:</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="sec-startdata.html#cb269-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(rna, <span class="at">file =</span> <span class="st">"data_output/my_rnaseq.csv"</span>)</span></code></pre></div>
<p>This new csv file can now be shared with other collaborators who
aren’t familiar with R.</p>
</div>
<div id="saving-data" class="section level3 unnumbered">
<h3>Saving data<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('saving-data')" onmouseout="reset_tooltip('saving-data-tooltip')"><span class="tooltiptext" id="saving-data-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>Exporting data to a spreadsheet has several limitations, such as those
described in the first chapter such as possible inconsistencies with
<code>,</code> and <code>.</code> for decimal separators and lack of variable type
definitions. Furthermore, exporting data to a spreadsheet is only
relevant for rectangular data such as dataframes and matrices.</p>
<p>A more general way to save data, that is specific to R and is
guaranteed to work on any operating system, is to use the <code>save</code>
function. Saving objects will generate a binary representation of the
object on disk, a <em>R Data</em> file (<code>rda</code> extension) that guarantees to
produce the same object once loaded back into R using the <code>load</code>
function.</p>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="sec-startdata.html#cb270-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(rna, <span class="at">file =</span> <span class="st">"data_output/rnaseq.rda"</span>)</span>
<span id="cb270-2"><a href="sec-startdata.html#cb270-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(rna)</span>
<span id="cb270-3"><a href="sec-startdata.html#cb270-3" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"data_output/rnaseq.rda"</span>)</span>
<span id="cb270-4"><a href="sec-startdata.html#cb270-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(rna)</span></code></pre></div>
<p>Note about how the function <code>load</code> loads the object in the file
directly in the global environment.</p>
<p>There is also the <code>saveRDS</code> and <code>readRDS</code> functions that save R
objects to binary files (using the <code>rds</code> extension here) and read
these back into R. From a user’s perspective, main different is that,
<code>load</code> loads an object in the global environment while <code>readRDS</code> reads
the data from disk and returns it. It is this necessary to store the
output of <code>readRDS</code> in a variable:</p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb271-1"><a href="sec-startdata.html#cb271-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(rna, <span class="at">file =</span> <span class="st">"data_output/rnaseq.rds"</span>)</span>
<span id="cb271-2"><a href="sec-startdata.html#cb271-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(rna)</span>
<span id="cb271-3"><a href="sec-startdata.html#cb271-3" aria-hidden="true" tabindex="-1"></a>rna <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data_output/rnaseq.rds"</span>)</span>
<span id="cb271-4"><a href="sec-startdata.html#cb271-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(rna)</span></code></pre></div>
<pre><code>##      gene     sample expression     organism age    sex  infection  strain time
## 1     Asl GSM2545336       1170 Mus musculus   8 Female InfluenzaA C57BL/6    8
## 2    Apod GSM2545336      36194 Mus musculus   8 Female InfluenzaA C57BL/6    8
## 3 Cyp2d22 GSM2545336       4060 Mus musculus   8 Female InfluenzaA C57BL/6    8
## 4    Klk6 GSM2545336        287 Mus musculus   8 Female InfluenzaA C57BL/6    8
## 5   Fcrls GSM2545336         85 Mus musculus   8 Female InfluenzaA C57BL/6    8
## 6  Slc2a4 GSM2545336        782 Mus musculus   8 Female InfluenzaA C57BL/6    8
##       tissue mouse ENTREZID
## 1 Cerebellum    14   109900
## 2 Cerebellum    14    11815
## 3 Cerebellum    14    56448
## 4 Cerebellum    14    19144
## 5 Cerebellum    14    80891
## 6 Cerebellum    14    20528
##                                                                    product
## 1                            argininosuccinate lyase transcript variant X1
## 2                                    apolipoprotein D transcript variant 3
## 3 cytochrome P450 family 2 subfamily d polypeptide 22 transcript variant 2
## 4                      kallikrein related-peptidase 6 transcript variant 2
## 5              Fc receptor-like S scavenger receptor transcript variant X1
## 6       solute carrier family 2 (facilitated glucose transporter) member 4
##      ensembl_gene_id external_synonym chromosome_name   gene_biotype
## 1 ENSMUSG00000025533    2510006M18Rik               5 protein_coding
## 2 ENSMUSG00000022548             &lt;NA&gt;              16 protein_coding
## 3 ENSMUSG00000061740             2D22              15 protein_coding
## 4 ENSMUSG00000050063             Bssp               7 protein_coding
## 5 ENSMUSG00000015852    2810439C17Rik               3 protein_coding
## 6 ENSMUSG00000018566           Glut-4              11 protein_coding
##                             phenotype_description
## 1           abnormal circulating amino acid level
## 2                      abnormal lipid homeostasis
## 3                        abnormal skin morphology
## 4                         abnormal cytokine level
## 5 decreased CD8-positive alpha-beta T cell number
## 6              abnormal circulating glucose level
##   hsapiens_homolog_associated_gene_name
## 1                                   ASL
## 2                                  APOD
## 3                                CYP2D6
## 4                                  KLK6
## 5                                 FCRL4
## 6                                SLC2A4</code></pre>
<p>To conclude, when it comes to saving data from R that will be loaded
again in R, saving and loading is the preferred approach. If tabular
data need to be shared with somebody that is not using R, then
exporting to a text-based spreadsheet is a good alternative.</p>

</div>
</div>
</div></body></html>

<p style="text-align: center;">
<a href="sec-startr.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-dplyr.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2021-09-24
 using 
R version 4.1.0 (2021-05-18)
</p>
</div>
</div>



</body>
</html>
